src/main/resources/META-INF/rewrite/micronaut3-to-4.yml

```diff
@@ -0,0 +1,373 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.Micronaut3to4Migration
+displayName: Migrate from Micronaut 3.x to 4.x
+description: This recipe will apply changes required for migrating from Micronaut 3 to Micronaut 4.
+recipeList:
+  - org.openrewrite.java.micronaut.UpdateBuildToJava17
+  - org.openrewrite.gradle.UpdateGradleWrapper:
+      version: 8.x
+  - org.openrewrite.java.micronaut.UpdateBuildToMicronaut4Version
+  - org.openrewrite.java.micronaut.UpdateMicronautPlatformBom
+  - org.openrewrite.java.micronaut.UpdateBuildPlugins
+  - org.openrewrite.java.micronaut.AddSnakeYamlDependencyIfNeeded
+  - org.openrewrite.java.micronaut.UpdateJakartaAnnotations
+  - org.openrewrite.java.micronaut.UpdateJakartaPersistence
+  - org.openrewrite.java.micronaut.UpdateMicronautValidation
+  - org.openrewrite.java.micronaut.UpdateMicronautEmail
+  - org.openrewrite.java.micronaut.AddMicronautWebsocketDependencyIfNeeded
+  - org.openrewrite.java.micronaut.UpdateMicronautSession
+  - org.openrewrite.java.micronaut.AddMicronautRetryDependencyIfNeeded
+  - org.openrewrite.java.micronaut.UpdateMicronautSecurity
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateBuildToJava17
+displayName: Update the Gradle or Maven build to a Java 17 baseline
+description: This recipe will update the Java version to 17 for a Gradle or Maven build.
+recipeList:
+  - org.openrewrite.gradle.UpdateJavaCompatibility:
+      version: 17
+  - org.openrewrite.maven.ChangePropertyValue:
+      key: jdk.version
+      newValue: 17
+  - org.openrewrite.maven.ChangePropertyValue:
+      key: release.version
+      newValue: 17
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateBuildToMicronaut4Version
+displayName: Update the Micronaut version to 4.x
+description: This recipe will update the Micronaut version to 4.x for a Gradle or Maven build.
+recipeList:
+  - org.openrewrite.java.micronaut.UpgradeMicronautMavenPropertyVersion:
+      newVersion: 4.0.0-M2
+  - org.openrewrite.java.micronaut.UpgradeMicronautGradlePropertiesVersion:
+      newVersion: 4.0.0-M2
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateBuildPlugins
+displayName: Add Micronaut build plugins to 4.x
+description: This recipe will update the shadow jar plugin to 8.x and the Micronaut build plugins to 4.x for a Gradle build.
+recipeList:
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: com.github.johnrengelman.shadow
+      newVersion: 8.x
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.aot
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.application
+      newVersion: 4.0.0-M2
```

Comment:
Yea, our versioning can be a bit confusing...but basically I'm trying to keep things aligned with what we generate at https://micronaut.io/launch - I will update this once we've released M3 of the full platform. Generally, it should be OK to have a newer version of the plugin vs what is generated by launch, so I will consider changing that to the Xrange selector in the future.

Will the Xrange selector in these general recipes work with milestone versions like this? Would 4.0.x result in finding 4.0.0-M4? 

I tried using the Xrange selector initially when I first started working on this, but I found the logic in `MicronautVersionHelper` didn't work when specifying 4.0.x because of the milestone tag, so I had to use the explicit version, and then that sent me down the path of doing that everywhere.

Reply:
You may be right that [our Xrange matcher](https://github.com/openrewrite/rewrite/blob/main/rewrite-core/src/main/java/org/openrewrite/semver/XRange.java) does not yet support milestones, [looking at the tests](https://github.com/openrewrite/rewrite/blob/main/rewrite-core/src/test/java/org/openrewrite/semver/XRangeTest.java). In that case we might use 4.0.0-M2 until 4.0.0 is available, and at that point switch over to 4.0.x, to ensure the tests don't break when 4.0.1 comes out.